packages/tpetra/core/src/Tpetra_Details_residual.hpp

```diff
@@ -0,0 +1,294 @@
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (X.getNumVectors () != R.getNumVectors (), std::runtime_error,
+       "X.getNumVectors() = " << X.getNumVectors () << " != "
+       "R.getNumVectors() = " << R.getNumVectors () << ".");
+
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (X.getLocalLength () !=
+       A.getColMap ()->getNodeNumElements (), std::runtime_error,
+       "X has the wrong number of local rows.  "
+       "X.getLocalLength() = " << X.getLocalLength () << " != "
+       "A.getColMap()->getNodeNumElements() = " <<
+       A.getColMap ()->getNodeNumElements () << ".");
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (R.getLocalLength () !=
+       A.getRowMap ()->getNodeNumElements (), std::runtime_error,
+       "R has the wrong number of local rows.  "
+       "R.getLocalLength() = " << R.getLocalLength () << " != "
+       "A.getRowMap()->getNodeNumElements() = " <<
+       A.getRowMap ()->getNodeNumElements () << ".");
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (B.getLocalLength () !=
+       A.getRowMap ()->getNodeNumElements (), std::runtime_error,
+       "B has the wrong number of local rows.  "
+       "B.getLocalLength() = " << B.getLocalLength () << " != "
+       "A.getRowMap()->getNodeNumElements() = " <<
+       A.getRowMap ()->getNodeNumElements () << ".");
+
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (! A.isFillComplete (), std::runtime_error, "The matrix A is not "
+       "fill complete.  You must call fillComplete() (possibly with "
+       "domain and range Map arguments) without an intervening "
+       "resumeFill() call before you may call this method.");
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (! X.isConstantStride () || ! R.isConstantStride () || ! B.isConstantStride (),
+       std::runtime_error, "X, Y and B must be constant stride.");
+    // If the two pointers are NULL, then they don't alias one
+    // another, even though they are equal.
+    TEUCHOS_TEST_FOR_EXCEPTION
+      ((X_lcl.data () == R_lcl.data () && X_lcl.data () != nullptr) ||
+       (X_lcl.data () == B_lcl.data () && X_lcl.data () != nullptr),
+       std::runtime_error, "X, Y and R may not alias one another.");
+  }
+      
+  // This is currently a "reference implementation" waiting until Kokkos Kernels provides
+  // a residual kernel.
+  SC one = Teuchos::ScalarTraits<SC>::one(), negone = -one, zero = Teuchos::ScalarTraits<SC>::zero();    
+  A.localApply(X,R,Teuchos::NO_TRANS, one, zero);
+  R.update(one,B,negone);
+}
+    
+
+//! Computes R = B - A * X 
+template<class SC, class LO, class GO, class NO>
+void residual(const Operator<SC,LO,GO,NO> &   Aop,
+              const MultiVector<SC,LO,GO,NO> & X_in,
+              const MultiVector<SC,LO,GO,NO> & B_in,
+              MultiVector<SC,LO,GO,NO> & R_in) {
+  using Tpetra::Details::ProfilingRegion;
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+  using Teuchos::rcp_const_cast;
+  using Teuchos::rcpFromRef;
+  
+  const CrsMatrix<SC,LO,GO,NO> * Apt  = dynamic_cast<const CrsMatrix<SC,LO,GO,NO>*>(&Aop);
```

Comment:
I would prefer the opposite code organization:

  * The overload (above) that takes CrsMatrix should have the optimized implementation.
  * The overload here that takes Operator should first attempt the `dynamic_cast`, if it succeeds, it should just invoke the CrsMatrix overload; otherwise, it should have a fall-back implementation that works for Operator.

This organization would mean that you only need to implement the optimized version once.

Reply:
Not entirely following --- I only have one optimized implementation (OK, it isn't optimized, now, but it will be).

Reply:
Here's what I mean:

1. Overload of `residual` that takes `CrsMatrix` does the optimized thing.
2. Overload of `residual` that takes `Operator` first attempts the `dynamic_cast` to CrsMatrix.  If the cast succeeds, then it should just call the overload of `residual` that takes `CrsMatrix`.  If not, then it should do the generic Operator thing.